File: Examples/android/CMeSdkJava/app_rework/src/main/java/me/connect/sdk/java/sample/connections/ConnectionsViewModel.java
package me.connect.sdk.java.sample.connections;

import android.app.Application;
import android.net.Uri;
import android.util.Base64;
import android.webkit.URLUtil;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import org.json.JSONException;
import org.json.JSONObject;


import java.util.List;
import java.util.concurrent.Executors;

import me.connect.sdk.java.ConnectMeVcxUpdated;
import me.connect.sdk.java.connection.QRConnection;
import me.connect.sdk.java.message.MessageState;
import me.connect.sdk.java.sample.SingleLiveData;
import me.connect.sdk.java.sample.db.Database;
import me.connect.sdk.java.sample.db.entity.Connection;

public class ConnectionsViewModel extends AndroidViewModel {
    private final Database db;
    private MutableLiveData<List<Connection>> connections;

    public ConnectionsViewModel(@NonNull Application application) {
        super(application);
        db = Database.getInstance(application);
    }

    public LiveData<List<Connection>> getConnections() {
        if (connections == null) {
            connections = new MutableLiveData<>();
        }
        loadConnections();
        return connections;
    }

    public SingleLiveData<Boolean> newConnection(String invite) {
        SingleLiveData<Boolean> data = new SingleLiveData<>();
        createConnection(invite, data);
        return data;
    }

    private void loadConnections() {
        Executors.newSingleThreadExecutor().execute(() -> {
            List<Connection> data = db.connectionDao().getAll();
            connections.postValue(data);
        });
    }

    private void createConnection(String invite, SingleLiveData<Boolean> liveData) {
        Executors.newSingleThreadExecutor().execute(() -> {
            String parsedInvite = parseInvite(invite);
            ConnDataHolder data = extractDataFromInvite(parsedInvite);
            ConnectMeVcxUpdated.createConnection(parsedInvite, new QRConnection())
                    .handle((res, throwable) -> {
                        if (res != null) {
                            String serializedCon = ConnectMeVcxUpdated.awaitConnectionStatusChange(res, MessageState.ACCEPTED);
                            Connection c = new Connection();
                            c.name = data.name;
                            c.icon = data.logo;
                            c.serialized = serializedCon;
                            db.connectionDao().insertAll(c);
                            loadConnections();
                        }
                        liveData.postValue(throwable == null);
                        return res;
                    });
        });
    }

    private String parseInvite(String invite) {
        if (URLUtil.isValidUrl(invite)) {
            Uri uri = Uri.parse(invite);
            String param = uri.getQueryParameter("c_i");
            if (param != null) {
                return new String(Base64.decode(param, Base64.NO_WRAP));
            } else {
                return "";
            }
        } else {
            return invite;
        }
    }

    private ConnDataHolder extractDataFromInvite(String invite) {
        try {
            JSONObject json = new JSONObject(invite);
            if (json.has("label")) {
                String label = json.getString("label");
                return new ConnDataHolder(label, null);
            }
            JSONObject data = json.optJSONObject("s");
            if (data != null) {
                return new ConnDataHolder(data.getString("n"), data.getString("l"));
            } else {
                // workaround in case details missing
                String sourceId = json.getString("id");
                return new ConnDataHolder(sourceId, null);
            }
        } catch (JSONException e) {
            e.printStackTrace();
            return null;
        }
    }

    static class ConnDataHolder {
        String name;
        String logo;

        public ConnDataHolder(String name, String logo) {
            this.name = name;
            this.logo = logo;
        }
    }
}

File: Examples/android/CMeSdkJava/app_rework_kt/src/main/kotlin/me/connect/sdk/java/samplekt/connections/ConnectionsViewModel.kt
package me.connect.sdk.java.samplekt.connections

import android.app.Application
import android.net.Uri
import android.util.Base64
import android.webkit.URLUtil
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.future.await
import kotlinx.coroutines.launch
import me.connect.sdk.java.samplekt.SingleLiveData
import me.connect.sdk.java.samplekt.db.Database
import me.connect.sdk.java.samplekt.db.entity.Connection
import me.connect.sdk.java.Connections
import me.connect.sdk.java.connection.QRConnection
import me.connect.sdk.java.message.MessageState
import me.connect.sdk.java.samplekt.wrap
import org.json.JSONException
import org.json.JSONObject
import java.lang.Exception


class ConnectionsViewModel(application: Application) : AndroidViewModel(application) {
    private val db: Database = Database.getInstance(application)
    private val connections: MutableLiveData<List<Connection>> by lazy {
        MutableLiveData<List<Connection>>()
    }

    fun getConnections(): LiveData<List<Connection>> {
        loadConnections()
        return connections
    }

    fun newConnection(invite: String): SingleLiveData<Boolean> {
        val data = SingleLiveData<Boolean>()
        createConnection(invite, data)
        return data
    }

    private fun loadConnections() = viewModelScope.launch(Dispatchers.IO) {
        val data = db.connectionDao().getAll()
        connections.postValue(data)
    }


    private fun createConnection(invite: String, liveData: SingleLiveData<Boolean>) = viewModelScope.launch(Dispatchers.IO) {
        try {
            val parsedInvite = parseInvite(invite)
            val data = extractDataFromInvite(parsedInvite)
            val res = Connections.create(parsedInvite, QRConnection()).wrap().await()
            val serializedCon = Connections.awaitStatusChange(res, MessageState.ACCEPTED)
            val c = Connection(
                    name = data!!.name,
                    icon = data.logo,
                    serialized = serializedCon
            )
            db.connectionDao().insertAll(c)
            loadConnections()
            liveData.postValue(true)
        } catch (e: Exception) {
            e.printStackTrace()
            liveData.postValue(false)
        }
    }

    private fun parseInvite(invite: String): String = if (URLUtil.isValidUrl(invite)) {
        val uri = Uri.parse(invite)
        val param = uri.getQueryParameter("c_i")
        if (param != null) {
            String(Base64.decode(param, Base64.NO_WRAP))
        } else {
            ""
        }
    } else {
        invite
    }

    private fun extractDataFromInvite(invite: String): ConnDataHolder? {
        return try {
            val json = JSONObject(invite)
            if (json.has("label")) {
                val label = json.getString("label")
                return ConnDataHolder(label, null)
            }
            val data = json.optJSONObject("s")
            if (data != null) {
                ConnDataHolder(data.getString("n"), data.getString("l"))
            } else {
                // workaround in case details missing
                val sourceId: String = json.getString("id")
                ConnDataHolder(sourceId, null)
            }
        } catch (e: JSONException) {
            e.printStackTrace()
            null
        }
    }

    internal class ConnDataHolder(var name: String, var logo: String?)
}
